### ISSUE TITLE
update_input and "vtkobjects" map documentation needed

### ISSUE THREAD

**Issue by dlonie** [UV-CDAT/uvcdat#1335]
@chaosphere2112 @doutriaux1 

I'm moving the "update_input" logic into the pipeline subclasses so that each plot's rendering can be encapsulated in a single spot.

I can port the functionality over rather easily, but need some help with documentation. Can one of you review the following and make corrections / fill in the blanks? Also, please document these things *as they're added* next time an approach like this is used. Trying to track down and reverse engineer each of these to figure out what they're doing was a royal PITA, and I'm still not sure about some of them. Explicit > implicit.

`VTKPlots.update_input(self, vtkobjects, array1, array2=None, update=True)`: Redraws a previous plot command using a new input dataset.
- `vtkobjects`: The map of objects returned from the call to `VTKPlots.plot`.
- `array1`: ??? (I know this is an input array, but what are the expected formats?)
- `array2`: ??? (Same as above, and when wouldn't it be `None`?)
- `update`: Force a render after updating the inputs if `True`. If `False`, only the inputs will be updated.

`vtk_backend_actors`: Is this just a list of all `vtkActor`s added to the renderwindow?

`vtk_backend_contours`: Instances of `vtkBandedPolyDataContourFilter`?

`vtk_backend_contours_labels_text_properties`: `vtkTextProperty` instances for labels isolines?

`vtk_backend_crdate_text_actor`: `vtkTextActor` instance with (plot/data)? creation date.

`vtk_backend_crtime_text_actor`: `vtkTextActor` instance with (plot/data)? creation time.

`vtk_backend_fillarea_actors`: Holds the result of undocumented function `vcs2vtk.prepFillarea`. Doesn't seem to be used anywhere.

`vtk_backend_filter`: Instance of `vtkDataSetSurfaceFilter` used to generate `vtkPolyData` from the input dataset.

`vtk_backend_geo`: Instance of `vtkGeoTransform` used to project (???) into (???) for the purpose of (???).

`vtk_backend_geofilters`: Instance of `vtkDataSetSurfaceFilter` used to convert `vtkThreshold` output from `vtkUnstructuredGrid` into `vtkPolyData` for rendering.

`vtk_backend_glyphfilters`: Instance of `vtkGlyph2D` used to generate glyphed polydata for vector plots.

`vtk_backend_grid`: The input dataset converted to a `vtkDataSet` subclass, either `vtkStructuredGrid` or `vtkUnstructuredGrid` (is this correct?).

`vtk_backend_labeled_luts`: The only surviving usage I can find is in isolinepipeline:

```
self._resultDict["vtk_backend_labeled_luts"] = [
                  [lut,
                   [self._contourLevels[0], self._contourLevels[-1], False]]]
```

I gather that `lut` is a `vtkLookupTable`, and the second item in the list is another list with `[lowContour, highContour, someOpaqueUndocumentedBoolean]`. I need some light shed on the boolean. Also, why is this needed in addition to `vtk_backend_luts`, which appears to store the same thing?

`vtk_backend_line_actors`: Stores the result of undocumented function `vcs2vtk.prepLine`. Appears to be unused.

`vtk_backend_luts`: See `vtk_backend_labeled_luts` for question about members and format.

`vtk_backend_marker_actors`: Stores the result of undocumented function `vcs2vtk.prepMarker`. Isn't used in `update_plot`, but is accessed by the interactive UI classes.

`vtk_backend_missing_mapper`: A tuple containing `(mapper, None, someBoolean)`. I gather that:
- `mapper`: The `vtkMapper` used to render masks over invalid regions of a dataset.
- `None`: ??? What is this, what can it be, why would it be used?
- `someBoolean`: ??? What is this, what is it used for?

`vtk_backend_Min_text_actor`: Instance of `vtkTextActor` that displays the minimum scalar value for the dataset.

`vtk_backend_Max_text_actor`: Instance of `vtkTextActor` that displays the maximum scalar value for the dataset.

`vtk_backend_Mean_text_actor`: Instance of `vtkTextActor` that displays the mean scalar value for the dataset.

`vtk_backend_text_actors`: List of `vtkTextActor` objects produced by `vcs2vtk.genTextActor` for `plot` calls with the `text` graphics method.

`vtk_backend_zvalue_text_actor`: Instance of `vtkTextActor` added in `VTKPlots.renderTemplate`, I assume it shows the z-value of a dataslice??? Doesn't seem to be used anywhere.

**Comment by chaosphere2112** [comment]
@dlonie `vtk_backend_actors` is, indeed, the list of actors added for that plot. It's used in the animation code and in the interactivity code.

**Comment by chaosphere2112** [comment]
@dlonie 
All of the `vtk_backend_*_text_actor` objects are labels from the template; `zvalue` is the current time, by default. `fillarea_actors` and `line_actors` will probably be used by interaction code when (if?) the time comes (I wrote the code to edit fillareas and lines, but Dean suggested that we trim that out for the present to focus on text and markers, which are the most commonly used items). If we've got the chance, I'd love to refactor the Min, Max, and Mean text actors to just be lower case like the rest of them (I have a dict that I use to convert between the label name and the text actor key just for those three), but I could see that being a huge pain.

`prepMarker` outputs:

 - `vtk.vtkGlyph2D` (used to define the markers)
 - `vtk.vtkGlyphSource2D` (if the marker is of a glyph type supported by VTK, else None)
 - `vtk.vtkPolyData` (if the marker is not of a glyph type supported by VTK, else None)
 - `vtk.vtkActor` (actor used to display the markers)
 - `vtk.vtkGeoTransform` (result of `vcs2vtk.project`, which falls into the realm of "things that are too math-y/graphics-y for me to understand without spending a lot of time looking at them")

**Comment by dlonie** [comment]
@chaosphere2112 @doutriaux1 

I'm about to start adding API to the pipeline objects to return the objects needed for the animation, interaction, etc. Which objects need to be exposed? Telling me the current key would be fine.

I think it would make sense to just have a "getPipelineObject(key, default=None)" method that the animation/interact code can use to get objects out of the pipeline. It'd save us from having to add pipeline-specific API across all of the pipelines, even when they don't make sense. It'd be similar to the current dictionary-based approach, but would let us manage the lifetimes of these objects more easily.

For reference, here are all instances of "vtk_backend_..." that appear outside of `update_input` and the pipeline objects. Are all of these needed? Or just certain ones?

```
Packages/vcs/Lib/VTKPlots.py:482:        returned["vtk_backend_text_actors"] = vcs2vtk.genTextActor(ren,to=to,tt=tt)
Packages/vcs/Lib/VTKPlots.py:487:        returned["vtk_backend_line_actors"]=actors
Packages/vcs/Lib/VTKPlots.py:493:        returned["vtk_backend_marker_actors"]=actors
Packages/vcs/Lib/VTKPlots.py:502:        returned["vtk_backend_fillarea_actors"]=actors
Packages/vcs/Lib/VTKPlots.py:578:        texts=d.backend.get("vtk_backend_text_actors",[])
Packages/vcs/Lib/VTKPlots.py:584:              returned["vtk_backend_%s_text_actor" % s0] = t
Packages/vcs/Lib/VTKPlots.py:586:              returned["vtk_backend_%s_text_actor" % d.backend["vtk_backend_template_attribute"]] = t
Packages/vcs/Lib/VTKPlots.py:616:            returned["vtk_backend_crdate_text_actor"]=actors[0]
Packages/vcs/Lib/VTKPlots.py:625:            returned["vtk_backend_crtime_text_actor"]=actors[0]
Packages/vcs/Lib/VTKPlots.py:672:            returned["vtk_backend_zvalue_text_actor"]=actors[0]

Packages/vcs/Lib/configurator.py:679:            key = "vtk_backend_text_actors"
Packages/vcs/Lib/configurator.py:687:            key = "vtk_backend_marker_actors"

Packages/vcs/Lib/editors/label.py:17:    "mean": "vtk_backend_Mean_text_actor",
Packages/vcs/Lib/editors/label.py:18:    "min": "vtk_backend_Min_text_actor",
Packages/vcs/Lib/editors/label.py:19:    "max": "vtk_backend_Max_text_actor",
Packages/vcs/Lib/editors/label.py:26:    elif "vtk_backend_%s_text_actor" % member.member in dp.backend:
Packages/vcs/Lib/editors/label.py:27:        actor = dp.backend["vtk_backend_%s_text_actor" % member.member]

Packages/vcs/Lib/editors/marker.py:22:        actors = display.backend["vtk_backend_marker_actors"][index]
Packages/vcs/Lib/editors/marker.py:97:        self.display.backend["vtk_backend_marker_actors"][self.index] = (self.glyph, self.glyph_source, self.polydata, self.actor, self.geo)

Packages/vcs/Lib/editors/text.py:33:        self.actors = dp.backend["vtk_backend_text_actors"]

Packages/vcs/Lib/template.py:1318:                    dp.backend["vtk_backend_template_attribute"] = s
Packages/vcs/Lib/template.py:1320:                    dp.backend["vtk_backend_template_attribute"] = "dataname"

Packages/vcs/Lib/vcs2vtk.py:164:  out={"vtk_backend_grid":vg,
Packages/vcs/Lib/vcs2vtk.py:355:  out={"vtk_backend_grid":vg,
```

My plan at this point is to expose these in the pipeline API, then let you guys port your code over to use the new API.

**Comment by doutriaux1** [comment]
thanks @dlonie the idea is that this vtk_backend were all the stuff we need to update either when going to the next time step in animation (which basically you're replacing with exposing the pipeline object). 
The other use of these object was for "ui interaction", it was a FIRST set VERY limited. Eventually we need to be able to access/move/edit EVERY object controlled by the "template". i.e all text objects on the plot, their font/location/string, and the "legend" and "data" areas (i.e. their size/location).

I'm not usre how best to achieve this with VTK but I'm guessing some sort of template pipeline is needed. 

Basically the graphic methods (boxfillpipeline etc...) describe how to plot things (colors/levels/type of plot, etc...) and the template objects define WHERE to plot them (and how for text object such as labels)

Hope this helps rather than confuses you.

**Comment by dlonie** [comment]
Ok, for now I'll just add a generic key-based retrieval API to `Pipeline` that can be used to get out any items you may need.
